### PR TITLE
[REVIEW] BFS add C++ tests

### LINKED ISSUE #778: [ENH] BFS - Test
**Describe the solution you'd like**
- [ ] C++ tests for BFS
- [ ] Tests should include predecessors validation for BFS in C++/ Python (BFS is tested in SSSP tests)

### PR DESCRIPTION
**Pull request by xcadet**
BFS was tested through BC on the C++ side.
This adds a separate file for BFS testing at the C++ level.

In order to follow the same architecture as the `cugprah/cpp/src/` folder, it adds a `cugraph/cpp/tests/traversal/` folder containing `bfs_test.cu` and `sssp_test.cu`  (previously located in `cugraph/cpp/tests/sssp/`)

Added `generate_graph_csr_from_mm` in `cugraph/cpp/tests/tests_utils.h`

**To do in this PR:**
- [x] Verify distances.
- [x] Verify predecessor. (At this point cugpaphs' BFS only returns one predecessor, verify that it is a valid one)
- [x] Verify shortest path counting. (Component required in Betweenness Centrality)
- [x] Update compare code on python side.

Closes #778

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by xcadet** [comment]
rerun tests

File: python/cugraph/tests/test_bfs.py [hunk 1]
```diff
@@ -190,36 +191,50 @@ def _compare_bfs(G,  Gnx, source):
...
+
+    # cugraph return a dataframe that should contain exactly one time each
+    # vertex
+    # We could us isin to filter only vertices that are common to both
+    # But it would slow down the comparison, and in this specific case
+    # nxacb._single_source_shortest_path_basic is a dictionary containing all
+    # the vertices.
+    # There is no guarantee when we get `df` that the vertices are sorted
+    # thus we enforce the order so that we can leverage faster comparison after
+    sorted_df = df.sort_values('vertex').rename({"sp_counter": "cu_spc"})
```
Comment (xcadet) [inline python/cugraph/tests/test_bfs.py:213]: This speeds up the verification phase as soon as graphs have more than a few hundreds vertices:
- |V| = 33  :  0.0030s from 0.0010s
- |V| = 105:  0.0030s from 0.0045s
- |V| = 4k  :  0.0030s from 0.1870s
- |V| = 50M: 25.000s  from 2160.0s
We can check values that are present in both by filtering the DataFrame using `isin` which slows down the process a bit.
- + 0.004s for small size graphs
- + 3.000s for the |V| = 50M

Comment (xcadet) [reply python/cugraph/tests/test_bfs.py:213]: rerun tests

**Comment by xcadet** [comment]
rerun tests

**Pull request merged**